lib/CArtHandler.cpp

```diff
@@ -301,22 +303,42 @@ CArtHandler::~CArtHandler() = default;
 std::vector<JsonNode> CArtHandler::loadLegacyData()
 {
 	size_t dataSize = VLC->settings()->getInteger(EGameSettings::TEXTS_ARTIFACT);
-
 	objects.resize(dataSize);
+
 	std::vector<JsonNode> h3Data;
 	h3Data.reserve(dataSize);
 
 	#define ART_POS(x) #x ,
 	const std::vector<std::string> artSlots = { ART_POS_LIST };
 	#undef ART_POS
+	
+	std::set<std::string> artSlotAllowed;
 
 	static std::map<char, std::string> classes =
 		{{'S',"SPECIAL"}, {'T',"TREASURE"},{'N',"MINOR"},{'J',"MAJOR"},{'R',"RELIC"},};
 
 	CLegacyConfigParser parser("DATA/ARTRAITS.TXT");
 	CLegacyConfigParser events("DATA/ARTEVENT.TXT");
+	
+	parser.endLine();
+	
+	//parse header
+	for(auto slotName = parser.readString(); !slotName.empty(); slotName = parser.readString())
+	{
+		//try to match our format
+		for(auto & c : slotName)
+		{
+			c = std::toupper(c);
+			if(c == ' ')
+				c = '_';
+		}
+		
+		artSlotAllowed.insert(slotName);
+		slotName.erase(std::remove(slotName.begin(), slotName.end(), '_'), slotName.end()); //no spaces and underscores
+		slotName = std::regex_replace(slotName, std::regex("WARMACHINE"), "MACH"); //special case

```

Comment:
why regex? We always use boost:replace_first or boost::replace_all in such cases

Reply:
I prefer std. Actually, I was thinking that we could get rid of boost dependency at one day

Reply:
In that case should be done more consistently instead of spreading multiple different methods through code.

Also have concerns about regex performance compared to simple string replacement

Reply:
Don't see any complains about regex perfomance, so presumably acceptable.

If we're replacing boost::replace_first / boost::replace_all then I suggest to
- create vstd::replace_first / replace_first methods that will internally use std::regex
- replace all instances of boost usages with vstd

But please, our text processing is already a mess with massive mix of std::string operations, boost format, boost replace and who knows what else. No need to add regex to this pile.

Reply:
rather create regex overload. For simple text replacement you can use implementation from cppref: https://en.cppreference.com/w/cpp/string/basic_string/replace (in Example)

Reply:
Yeah, regex-free implementation with only std::string will probably be the best option.

That being said, my main concern is to avoid increasing mess in our already messy text manipulation code. If we want to replace boost then I suggest making small wrapper in vstd to cover all our typical use-cases and replace boost with that wrapper. But not introducing yet another way to manipulate strings.

Reply:
> I suggest making small wrapper in vstd to cover all our typical use-cases and replace boost with that wrapper

exactly my thoughts

Reply:
guys, it's short, clear and single line of code. it's not bug-prone and doesn't require refactoring. It doesn't use any thirdparty and is part of standard.

Please, explain why alternative solutions suggested are better and so important to use in vcmi to solve substring replacement problem.

Reply:
Consistency?
It is already tiring to figure out how each formatting operation works in vcmi.
"Oh, we use boost format here? Okay, then we need to wrap this in boost::str and use '%' operator to format stuff, and replaced string will be available as return type"
"Oh, but here we use boost replace, so replacement is done in place, and we need to specify string to be replaced"
"Oh, here we use simple concatenation of multiple std::strings, so simply use operator +"
And now also:
"Oh and here we use std regex so ...  where will be result and how to provide parameters?"

Please, VCMI codebase is already a mess and pretty much every area needs refactoring. Please, don't make this even more messy.

Right now we (mostly) use boost::replace_first. So please, keep using it or refactor code to use better routine.

Reply:
ok, I see there is no a single usage of regex in codebase. Refactoring is definitely out of this review